File: algorithms/golang/205.同构字符串.go
/*
 * @lc app=leetcode.cn id=205 lang=golang
 *
 * [205] 同构字符串
 */

// @lc code=start
func isIsomorphic(s string, t string) bool {
	sLen, tLen := len(s), len(t)
	if sLen != tLen {
		return false
	}
	cache, used := [256]byte{}, [256]bool{}
	for i := 0; i < sLen; i++ {
		ch1, ch2 := s[i], t[i]
		if cache[int(ch1)] == 0 && used[int(ch2)] || cache[int(ch1)] != 0 && cache[int(ch1)] != ch2 {
			return false
		}
		cache[int(ch1)] = ch2
		used[int(ch2)] = true
	}
	return true
}

// @lc code=end

File: algorithms/js/205.同构字符串.js
/*
 * @lc app=leetcode.cn id=205 lang=javascript
 *
 * [205] 同构字符串
 */

// @lc code=start
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isIsomorphic = function(s, t) {
  if (s.length !== t.length) {
    return false;
  }
  const hash = {},
    used = new Array(256);
  for (let i = 0; i < s.length; i++) {
    if (
      (!hash[s[i]] && used[t.charCodeAt(i)]) ||
      (hash[s[i]] && hash[s[i]] != t[i])
    ) {
      return false;
    }
    hash[s[i]] = t[i];
    used[t.charCodeAt(i)] = true;
  }
  return true;
};
// @lc code=end
